ocaml/typing/typemode.mli

```diff
@@ -12,13 +12,14 @@ val transl_alloc_mode : Parsetree.modes -> Mode.Alloc.Const.t
     requires different levels of maturity. *)
 val transl_modalities :
   maturity:Language_extension.maturity ->
-  has_mutable_implied_modalities:bool ->
+  has_mutable_implied_modalities:(bool, bool) Mode.monadic_comonadic ->
```

Comment:
I find this to be a slightly awkward interface. In the two places where `has_mutable_implied_modalities` have to get constructed (in `Printtyp.tree_of_label` and `Typedecl.transl_labels`), the process of construction is the same: check the mutability flag and then look for the attribute. Given that `transl_modalities` is responsible for understanding a list of modalities, I would expect it to take the mutability of the label and the attributes and then do the right thing around translation. Or maybe really we want a

```
val get_implied_modalities : mutable_:bool -> attributes -> modality list
```

which could get used in both `Typedecl` and `Printtyp`, I think.

Would that work? Might it simplify some things? In particular, I don't think the monadic/comonadic distinction works so great at usage sites: both usages of the implied modalities just lump together the comonadic ones and the monadic ones.

Reply:
Yes I like this idea.

Reply:
In the end I did something else (that I think is better). The key point is that, in general, the order in which modalities are applied matters. Mutable-implied modalities should preceed explicit modalities, and there should be a single place where this order is defined, and that's `typemode.ml`.